src/dns/LookupDetails.cc

```diff
@@ -16,8 +16,8 @@ Dns::LookupDetails::print(std::ostream &os) const
 {
     if (wait > 0)
         os << "lookup_wait=" << wait;
-    if (error.size())
-        os << " lookup_err=" << error;
+    if (error && error->length())
```

Comment:
This and similar .length() changes look like red flags to me: Once we have std::optional, we should stop checking error length. Either there is an error or there is not. Two states; one condition (provided by std::optional).

I have not checked everything, but, most likely, you need to refactor LookupDetails constructor to ensure the following invariant: A present error is never empty.

Reply:
I agree with you, in theory. But we don't necessarily know - some initializations are forwarded (e.g. IpCacheLookupForwarder::forwardLookup) and might be passing nullptrs or empty strings. And enforcing such an invariant would only be possible at runtime, and it would constitute an API change.
So my strategy here is to keep on supporting the current API promises, and once we are further along in getting rid of char* things when things will become clearer.

Reply:
Alex, This addition of complexity is the direct consequence of requiring std::optional (trilean object) to be used to represent a boolean state. Callers and users now need to explicitly differentiate states {never-set OR nil} from {currently-set AND non-nil}.
The original API used the pointer (SBuf is a decorated smart pointer) being set or non-set (aka 0 length/size). Which is categorically simpler.

Reply:
> This addition of complexity is the direct consequence of requiring std::optional

Where do you see additional complexity (in this change request scope)? Here are the sketches for all three options:

Official code:

```C++
if (error.size())
    os << " lookup_err=" << error;
```

With std::optional:

```C++
if (error)
    os << " lookup_err=" << *error;
```

With SBuf:

```C++
if (error.length())
    os << " lookup_err=" << error;
```

I do not see any additional complexity. In fact, one could argue that std::optional makes this code a bit simpler or more readable. Additional complexity does exist in the current PR, but that code should be fixed. This change request is about that fix.

Reply:
The complexity is that you are wrong about this case (for now, with incomplete transition):

> With std::optional:
> 
> ```c++
> if (error)
>     os << " lookup_err=" << *error;
> ```
> 

With creators of LookupDetails initializing the ```error``` member to empty string on non-error responses the recipients if-statement has to be in the form @kinkie used.

Reply:
@yadij, I believe you are reaching incorrect conclusions because you are looking at (unfinished) PR code. I am looking at how the code will look like when this PR code is fixed.

@kinkie, this is exactly why I have recommended finishing this work _before_ asking Amos to re-review (or concluding that there is some real deadlock here).